### DISCUSSION TITLE
Any way to tell if a set of SHAs are reachable from HEAD or a specific branch?

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#4154] [Q&A]
In GitLens, we are using `git log --all --date-order` to build the Graph visualization, but since we haven't been able to find a way to get `git log` to include all stashes we first call `git stash list` to get all the stash SHAs, and then use stdin to pipe them into the `git log --all --date-order --stdin` call. This works well for that usage, though it would be great to not have to pipe them in.

But when we want to filter the Graph to a single branch (or HEAD), we can't just pipe the same stash SHAs into the `git log` call, as it will then pull in any commits required to make those stash SHAs "reachable". 

Is there a way (without looping over each SHA) to validate if those SHAs are reachable from a specific branch or HEAD? Or is there a way to get `git log` to not include them if they are not reachable?

Thanks!

**Comment by dscho** [comment]
My immediate reaction is to pipe the SHAs to `git rev-list ^HEAD --stdin` to find out which ones are _not_ reachable. Would that help?

**Reply by dscho** [reply]
This command will show _many_ SHAs, of course. At first I thought that `--no-walk` could suppress the unneeded ones, but that would probably not work because it would not suppress the reachable ones.

**Comment by dscho** [comment]
Another option I considered was using [`git merge-base --all HEAD <SHA>...`](https://git-scm.com/docs/git-merge-base), but that is subject to command-line length limits, as it does not have any `--stdin` option (yet). On the other hand, it might be a good solution if you call the command as often as needed: you just have to build the intersection of SHAs in the input and output to find the reachable ones.

**Reply by dscho** [reply]
I am a bit confused, though, as no stash revision will be reachable from `HEAD`, by virtue of its construction.

**Reply by dscho** [reply]
Also note that the `merge-base` method will accidentally exclude input SHAs from the output if another input SHA is reachable from them.
